reasoner/planner/AnswerCountEstimator.java

```diff
@@ -0,0 +1,507 @@
+                        long newEstimate = estimate.answerEstimate(set(v));
+                        if (newEstimate > existingEstimate) { // Biggest one serves as baseline.
+                            newUnaryEstimateCover.put(v, estimate);
+                        }
+                    });
+
+            return newUnaryEstimateCover;
+        }
+
+        private LocalEstimate deriveEstimateFromConcludable(Concludable concludable) {
+            if (concludable.isHas()) {
+                return answerCountModel.estimatesFromHasEdges(concludable.asHas());
+            } else if (concludable.isRelation()) {
+                return answerCountModel.estimatesFromRolePlayersAssumingEvenDistribution(concludable.asRelation());
+            } else if (concludable.isIsa()) {
+                assert concludable.generating().isPresent();
+                ThingVariable v = concludable.generating().get();
+                return new LocalEstimate.SimpleEstimate(list(v),
+                        answerCountModel.countPersistedThingsMatchingType(v) + answerCountModel.estimateInferredAnswerCount(concludable, set(v)));
+            } else throw TypeDBException.of(ILLEGAL_STATE);
+        }
+
+        private static abstract class LocalEstimate {
+
+            final List<Variable> variables;
+
+            private LocalEstimate(List<Variable> variables) {
+                this.variables = variables;
+            }
+
+            abstract long answerEstimate(Set<Variable> variableFilter);
+
+            private static class SimpleEstimate extends LocalEstimate {
+                private final long staticEstimate;
+
+                private SimpleEstimate(List<Variable> variables, long estimate) {
+                    super(variables);
+                    this.staticEstimate = estimate;
+                }
+
+                @Override
+                long answerEstimate(Set<Variable> variableFilter) {
+                    return staticEstimate;
+                }
+            }
+
+            public static class CoPlayerEstimate extends LocalEstimate {
+
+                private final Map<TypeVariable, Integer> rolePlayerCounts;
+                private final Map<TypeVariable, Long> rolePlayerEstimates;
+                private final double relationTypeEstimate;
+                private final long inferredRelationEstimate;
+
+                public CoPlayerEstimate(List<Variable> variables, double relationTypeEstimate,
+                                        Map<TypeVariable, Long> rolePlayerEstimates, Map<TypeVariable, Integer> rolePlayerCounts,
+                                        long inferredRelationEstimate) {
+                    super(variables);
+                    this.relationTypeEstimate = relationTypeEstimate;
+                    this.rolePlayerEstimates = rolePlayerEstimates;
+                    this.rolePlayerCounts = rolePlayerCounts;
+                    this.inferredRelationEstimate = inferredRelationEstimate;
+                }
```

Comment:
Store just enough information to be able to reasonably answer queries on subsets of role-players.
Since we can't accurately estimate the co-constraints between role-players in inferred relations (without propagating constraints from triggered rules), that's an approximation.